File: apps/briar/ios/BriarCompanion/App/ChannelAlertPresentation.swift
import Foundation

enum ChannelAlertTone: String, Equatable, Sendable {
    case error
    case warning
}

enum ChannelAlertPresentation {
    private static let errorKeyword = try! NSRegularExpression(
        pattern: #"\b(?:error|errors|exception|fatal|panic|critical|fail(?:ed|ure)?|traceback|alarm|alert)\b|에러|오류|실패|알람|치명|错误|失败|告警|报警|致命"#,
        options: [.caseInsensitive]
    )
    private static let warningKeyword = try! NSRegularExpression(
        pattern: #"\b(?:warn(?:ing)?|degraded)\b|경고|警告"#,
        options: [.caseInsensitive]
    )
    private static let stackTrace = try! NSRegularExpression(
        pattern: #"traceback \(most recent call last\)|^\s*at \S+|^\s*File "[^"]+", line \d+"#,
        options: [.caseInsensitive, .anchorsMatchLines]
    )

    static func prettyJSON(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.first == "{" || trimmed.first == "[" else { return nil }
        guard let data = trimmed.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data),
              JSONSerialization.isValidJSONObject(object),
              let pretty = try? JSONSerialization.data(
                withJSONObject: object,
                options: [.prettyPrinted]
              ),
              let text = String(data: pretty, encoding: .utf8)
        else { return nil }
        return text
    }

    static func formattedDump(_ value: String) -> String {
        prettyJSON(value) ?? value
    }

    static func isStackTrace(_ value: String) -> Bool {
        matches(stackTrace, value)
    }

    static func shouldCollapse(_ value: String, expand: Bool = false) -> Bool {
        if expand { return true }
        let lines = value.split(separator: "\n", omittingEmptySubsequences: false).count
        if prettyJSON(value) != nil && (lines > 2 || value.count > 120) { return true }
        if isStackTrace(value) { return true }
        return lines >= 8 || value.count >= 480
    }

    static func preview(
        _ value: String,
        maxLines: Int = 4,
        maxChars: Int = 280
    ) -> (preview: String, collapsed: Bool) {
        guard shouldCollapse(value) else { return (value, false) }
        let lines = value.split(separator: "\n", omittingEmptySubsequences: false)
        var preview = lines.prefix(maxLines).joined(separator: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if preview.count > maxChars {
            preview = String(preview.prefix(maxChars)).trimmingCharacters(in: .whitespacesAndNewlines) + "…"
        } else if lines.count > maxLines || value.count > preview.count {
            preview += "…"
        }
        return (preview, true)
    }

    static func tone(from text: String, authorType: ChannelMessage.Author.Kind? = nil) -> ChannelAlertTone? {
        if let jsonTone = jsonTone(text) { return jsonTone }
        if isStackTrace(text) { return .error }
        guard let keyword = keywordTone(text) else { return nil }
        if authorType == .webhook { return keyword }
        if prettyJSON(text) != nil || shouldCollapse(text) { return keyword }
        return nil
    }

    static func plainText(from blocks: [ChannelMessageBlock]) -> String {
        blocks.compactMap { block in
            switch block.type {
            case .header, .section:
                return block.textObject?.text
            case .markdown:
                return block.markdownText
            case .context:
                return (block.contextElements ?? []).map(\.text).joined(separator: " ")
            case .richText:
                return (block.richTextElements ?? []).map(richPlain).joined(separator: "\n")
            case .divider:
                return nil
            }
        }.joined(separator: "\n")
    }

    static func tone(from message: ChannelMessage) -> ChannelAlertTone? {
        let body: String
        if let blocks = message.blocks, !blocks.isEmpty {
            body = plainText(from: blocks)
        } else {
            body = message.body
        }
        return tone(from: "\(message.author.name)\n\(body)", authorType: message.author.type)
    }

    static func localizedReplyError(
        _ error: String?,
        locale: CompanionLocale
    ) -> String {
        if error == "No available Worker can run this Agent." {
            return L10n.text("이 Agent를 실행할 수 있는 사용 가능한 Worker가 없습니다.", locale: locale)
        }
        return error ?? L10n.text("실행이 실패했습니다", locale: locale)
    }

    private static func jsonTone(_ value: String) -> ChannelAlertTone? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.first == "{" || trimmed.first == "[" else { return nil }
        guard let data = trimmed.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data)
        else { return nil }
        return tone(fromJSON: object)
    }

    private static func tone(fromJSON value: Any) -> ChannelAlertTone? {
        if let array = value as? [Any] {
            return array.compactMap(tone(fromJSON:)).first
        }
        guard let record = value as? [String: Any] else { return nil }
        for key in ["level", "severity", "status", "error", "type", "kind"] {
            if let field = record[key] as? String, let tone = keywordTone(field) {
                return tone
            }
            if record[key] is [String: Any] { return .error }
        }
        if record["error"] != nil || record["exception"] != nil { return .error }
        return nil
    }

    private static func keywordTone(_ value: String) -> ChannelAlertTone? {
        if matches(errorKeyword, value) { return .error }
        if matches(warningKeyword, value) { return .warning }
        return nil
    }

    private static func matches(_ expression: NSRegularExpression, _ value: String) -> Bool {
        expression.firstMatch(
            in: value,
            range: NSRange(value.startIndex..<value.endIndex, in: value)
        ) != nil
    }

    private static func richPlain(_ element: ChannelRichTextElement) -> String {
        switch element.type {
        case .section, .quote, .preformatted:
            return (element.elements ?? []).map { inline in
                switch inline.type {
                case .text: inline.text ?? ""
                case .link: inline.text ?? inline.url ?? ""
                case .emoji: ":\(inline.name ?? "emoji"):"
                }
            }.joined()
        case .list:
            return (element.sections ?? []).map { section in
                section.elements.map { inline in
                    inline.text ?? inline.url ?? ":\(inline.name ?? "emoji"):"
                }.joined()
            }.joined(separator: "\n")
        }
    }
}

File: apps/briar/src/lib/channel-alert-presentation.ts
import {
  channelMessageBlocksFallback,
  type ChannelMessage,
  type ChannelMessageBlock,
} from "./channels-contract";

export type ChannelAlertTone = "error" | "warning";

export type ChannelAlertPreview = {
  preview: string;
  collapsed: boolean;
};

const ERROR_KEYWORD =
  /\b(?:error|errors|exception|fatal|panic|critical|fail(?:ed|ure)?|traceback|alarm|alert)\b|에러|오류|실패|알람|치명|错误|失败|告警|报警|致命/iu;
const WARNING_KEYWORD =
  /\b(?:warn(?:ing)?|degraded)\b|경고|警告/iu;
const STACK_TRACE =
  /traceback \(most recent call last\)|^\s*at \S+|^\s*File "[^"]+", line \d+/imu;

export function prettyPrintJson(value: string): string | null {
  const trimmed = value.trim();
  if (!trimmed.startsWith("{") && !trimmed.startsWith("[")) return null;
  try {
    const parsed: unknown = JSON.parse(trimmed);
    if (parsed === null || typeof parsed !== "object") return null;
    return JSON.stringify(parsed, null, 2);
  } catch {
    return null;
  }
}

export function isStackTrace(value: string): boolean {
  return STACK_TRACE.test(value);
}

export function formattedChannelDump(value: string): string {
  return prettyPrintJson(value) ?? value;
}

export function shouldCollapseChannelText(
  value: string,
  expand = false,
): boolean {
  if (expand) return true;
  const lines = value.split("\n").length;
  if (prettyPrintJson(value) && (lines > 2 || value.length > 120)) return true;
  if (isStackTrace(value)) return true;
  return lines >= 8 || value.length >= 480;
}

export function channelAlertPreview(
  value: string,
  options: { maxLines?: number; maxChars?: number; force?: boolean } = {},
): ChannelAlertPreview {
  const maxLines = options.maxLines ?? 4;
  const maxChars = options.maxChars ?? 280;
  const collapsed = options.force === true || shouldCollapseChannelText(value);
  if (!collapsed) return { preview: value, collapsed: false };

  const lines = value.split("\n");
  let preview = lines.slice(0, maxLines).join("\n").trimEnd();
  if (preview.length > maxChars) {
    preview = `${preview.slice(0, maxChars).trimEnd()}…`;
  } else if (lines.length > maxLines || value.length > preview.length) {
    preview = `${preview}…`;
  }
  return { preview, collapsed: true };
}

function keywordTone(value: string): ChannelAlertTone | null {
  if (ERROR_KEYWORD.test(value)) return "error";
  if (WARNING_KEYWORD.test(value)) return "warning";
  return null;
}

function toneFromUnknown(value: unknown): ChannelAlertTone | null {
  if (Array.isArray(value)) {
    for (const item of value) {
      const tone = toneFromUnknown(item);
      if (tone) return tone;
    }
    return null;
  }
  if (!value || typeof value !== "object") return null;
  const record = value as Record<string, unknown>;
  for (const field of [
    record.level,
    record.severity,
    record.status,
    record.error,
    record.type,
    record.kind,
  ]) {
    if (typeof field === "string") {
      const tone = keywordTone(field);
      if (tone) return tone;
    }
    if (field && typeof field === "object") return "error";
  }
  if (record.error != null || record.exception != null) return "error";
  return null;
}

function jsonAlertTone(value: string): ChannelAlertTone | null {
  const trimmed = value.trim();
  if (!trimmed.startsWith("{") && !trimmed.startsWith("[")) return null;
  try {
    return toneFromUnknown(JSON.parse(trimmed) as unknown);
  } catch {
    return null;
  }
}

export function channelAlertToneFromText(
  value: string,
  authorType?: string,
): ChannelAlertTone | null {
  const jsonTone = jsonAlertTone(value);
  if (jsonTone) return jsonTone;
  if (isStackTrace(value)) return "error";
  const keyword = keywordTone(value);
  if (!keyword) return null;
  if (authorType === "webhook") return keyword;
  if (prettyPrintJson(value) || shouldCollapseChannelText(value)) return keyword;
  return null;
}

export function channelMessageAlertText(message: {
  author: { type: string; name: string };
  body: string;
  blocks?: ChannelMessageBlock[] | null;
}): string {
  const body = message.blocks?.length
    ? channelMessageBlocksFallback(message.blocks)
    : message.body;
  return [message.author.name, body].filter(Boolean).join("\n");
}

export function channelAlertToneFromMessage(
  message: Pick<ChannelMessage, "author" | "body"> & {
    blocks?: ChannelMessageBlock[] | null;
  },
): ChannelAlertTone | null {
  return channelAlertToneFromText(
    channelMessageAlertText(message),
    message.author.type,
  );
}
